keps/sig-instrumentation/20200415-cardinality-enforcement.md

```diff
@@ -0,0 +1,276 @@
+---
+title: Dynamic Cardinality Enforcement
+authors:
+  - "@logicalhan"
+owning-sig: sig-instrumentation
+participating-sigs:
+  - sig-instrumentation
+reviewers:
+  - todo
+approvers:
+  - todo
+editor: todo
+creation-date: 2020-04-15
+last-updated: 2020-04-15
+status: provisional
+---
+
+# Dynamic Cardinality Enforcement
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Graduation Criteria](#graduation-criteria)
+- [Post-Beta tasks](#post-beta-tasks)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+TLDR; metrics with unbounded dimensions can cause memory issues in the components they instrument.
+
+The simple solution to this problem is to say "don't do that". We (SIG instrumentation) have already done in our instrumentation guidelines, which specifically states that ["one should know a comprehensive list of all possible values for a label at instrumentation time."](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/instrumentation.md#dimensionality--cardinality).
```

Comment:
This definitely seems like more of a recommendation than a hard rule.  Even in components _we own_ (e.g. kube-state-metrics), there are labels with unbounded cardinality (e.g. `pod` or `namespace`).  If we really wanted to prevent unbounded cardinality, we would do so at compile-time and do another migration.  But I don't think that is actually what we want... We need to be able to use labels with unbounded cardinality in some places, but want an escape hatch for when we mess up.

Reply:
Definitely the node/pod stuff is exceptional. But you guys are doing manual GC for metrics no?  Manual cleanup of metrics doesn't really feel to me like the general use-case.

Reply:
Ah, got it.  So if we can definitively make the statement that all calls to `metrics.NewGaugeVec` (for example) should have labels with a specific set of values, should we start enforcing that all metrics have a whitelist at compile-time?  It would obviously be more intrusive than whats proposed here, but wouldn't that actually allow us to get to 100% bounded cardinality metrics?

Reply:
Yeah, but emphasis on the intrusive part. I don't know how realistically achievable that would be since it would mean auditing/updating every single metric in the codebase.

Reply:
got it, makes sense.

Reply:
FWIW kube-state-metrics is so special it doesn't even use the prometheus client anymore, but as already mentioned there are other exceptions to the rule like `kubelet_node_name`, which is practically bound to 1 series per node, but the value is free form in itself.